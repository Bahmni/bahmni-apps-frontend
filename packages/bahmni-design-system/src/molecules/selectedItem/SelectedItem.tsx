import { Close } from '@carbon/icons-react';
import { Button, Column, Grid } from '@carbon/react';
import classNames from 'classnames';
import React, { ReactNode } from 'react';
import styles from './styles/SelectedItem.module.scss';

/**
 * SelectedItem is a common component that displays content with a close button.
 * It's typically used for displaying selected items in a list or form context
 * where users can dismiss/remove the selection.
 */
export interface SelectedItemProps {
  /**
   * The content to display within the selected item container
   */
  children: ReactNode;

  /**
   * Function called when the close button is clicked
   */
  onClose: () => void;

  /**
   * Optional additional CSS class name to apply to the component
   */
  className?: string;

  /**
   * Optional data test id for testing
   */
  dataTestId?: string;
}

/**
 * A component for displaying selected items with a close button
 */
export const SelectedItem: React.FC<SelectedItemProps> = ({
  children,
  onClose,
  className,
  dataTestId = 'selected-item',
}) => {
  return (
    <Grid
      narrow
      fullWidth
      className={classNames(styles.grid, className)}
      data-testid={dataTestId}
    >
      <Column sm={3} md={7} lg={15} xlg={15} className={styles.children}>
        {children}
      </Column>
      <Column sm={1} md={1} lg={1} xlg={1} className={styles.close}>
        <Button
          id="close-selected-item"
          size="lg"
          hasIconOnly
          renderIcon={Close}
          iconDescription="Selected Item Close"
          aria-label="Close Selected Item"
          onClick={onClose}
          data-testid={`${dataTestId}-close-button`}
        />
      </Column>
    </Grid>
  );
};
export default SelectedItem;
