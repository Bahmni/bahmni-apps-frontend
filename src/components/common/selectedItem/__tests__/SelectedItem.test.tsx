import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import SelectedItem from '../SelectedItem';
import { axe, toHaveNoViolations } from 'jest-axe';

expect.extend(toHaveNoViolations);

describe('SelectedItem', () => {
  describe('Happy Paths', () => {
    it('should render children correctly', () => {
      // Arrange
      const testContent = 'Test content';
      const mockOnClose = jest.fn();

      // Act
      render(
        <SelectedItem onClose={mockOnClose}>
          <div>{testContent}</div>
        </SelectedItem>,
      );

      // Assert
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('should call onClose when close button is clicked', () => {
      // Arrange
      const mockOnClose = jest.fn();
      const testContent = 'Test content';

      // Act
      render(
        <SelectedItem onClose={mockOnClose}>
          <div>{testContent}</div>
        </SelectedItem>,
      );

      const closeButton = screen.getByLabelText('Close Selected Item');
      fireEvent.click(closeButton);

      // Assert
      expect(mockOnClose).toHaveBeenCalledTimes(1);
    });
  });

  describe('Accessibility', () => {
    it('should not have accessibility violations', async () => {
      // Arrange      // Arrange
      const mockOnClose = jest.fn();
      const testContent = 'Test content';

      const { container } = render(
        <SelectedItem onClose={mockOnClose}>
          <div>{testContent}</div>
        </SelectedItem>,
      );

      // Act & Assert
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});

describe('Snapshot', () => {
  it('should match snapshot', () => {
    // Arrange
    const mockOnClose = jest.fn();
    const testContent = 'Test content';

    // Act
    const { container } = render(
      <SelectedItem onClose={mockOnClose}>
        <div>{testContent}</div>
      </SelectedItem>,
    );

    // Assert
    expect(container).toMatchSnapshot();
  });
});
