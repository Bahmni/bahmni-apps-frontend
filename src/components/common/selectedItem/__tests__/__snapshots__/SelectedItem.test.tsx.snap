// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot should match snapshot 1`] = `
<div>
  <div
    class="cds--css-grid cds--css-grid--narrow cds--css-grid--full-width"
    data-testid="selected-item"
  >
    <div
      class="cds--sm:col-span-3 cds--md:col-span-7 cds--lg:col-span-15 cds--xlg:col-span-15 cds--css-grid-column"
    >
      <div>
        Test content
      </div>
    </div>
    <div
      class="cds--sm:col-span-1 cds--md:col-span-1 cds--lg:col-span-1 cds--xlg:col-span-1 cds--css-grid-column"
    >
      <span
        class="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--top cds--tooltip cds--icon-tooltip"
      >
        <div
          class="cds--tooltip-trigger__wrapper"
        >
          <button
            aria-label="Close Selected Item"
            aria-labelledby="tooltip-«rd»"
            class="cds--btn cds--btn--lg cds--layout--size-lg cds--btn--primary cds--btn--icon-only"
            data-testid="selected-item-close-button"
            id="close-selected-item"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17.4141 16L24 9.4141 22.5859 8 16 14.5859 9.4143 8 8 9.4141 14.5859 16 8 22.5859 9.4143 24 16 17.4141 22.5859 24 24 22.5859 17.4141 16z"
              />
            </svg>
          </button>
        </div>
        <span
          aria-hidden="true"
          class="cds--popover"
          id="tooltip-«rd»"
          role="tooltip"
        >
          <span
            class="cds--popover-content cds--tooltip-content"
          >
            Close
          </span>
          <span
            class="cds--popover-caret"
          />
        </span>
      </span>
    </div>
  </div>
</div>
`;
